patches/components-sync-driver-profile_sync_service.h.patch

```diff
@@ -1,7 +1,16 @@
 diff --git a/components/sync/driver/profile_sync_service.h b/components/sync/driver/profile_sync_service.h
-index 83a46645ee08815fe8f41da81d0a753480f8192e..cf720327932f3f2b300f1f5adb33b89d616ab6b0 100644
+index 83a46645ee08815fe8f41da81d0a753480f8192e..5bf53f77a7f65be1fef46d58b03886a92fd757ed 100644
 --- a/components/sync/driver/profile_sync_service.h
 +++ b/components/sync/driver/profile_sync_service.h
+@@ -106,7 +106,7 @@ class ProfileSyncService : public SyncService,
+   // Initializes the object. This must be called at most once, and
+   // immediately after an object of this class is constructed.
+   // TODO(mastiz): Rename this to Start().
+-  void Initialize();
++  void virtual Initialize();
```

Comment:
By reason that `Initialize` word is occurred an many headers, the trick of `#define Initialize virtual Initialize` didn't work. 
So I made a straight patch. Is there more good way?

Here is what happened:
```
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:160:22: error: 'virtual' can only appear on non-static member functions
  static inline void Initialize(Type** x, Arena* arena);
                     ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:214:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(STRING)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:215:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(BYTES)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:216:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(INT64)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:217:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(UINT64)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:218:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(INT32)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../chrome/browser/federated_learning/floc_id_provider_impl.cc:16:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:219:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(UINT32)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize

```

Reply:
Just to double check: when using an override in `chromium_src/components/sync/driver/profile_sync_service.h` to add a `#define Initialize virtual Initialize` line, did you by any chance forget to include also the #undef Initialize` line?

From your comment and the error above it seems as if you were missing that `#undef`, which I think would explain the build error. See for instance `chromium_src/chrome/browser/browser_process_impl.h`:
```
#define Init virtual Init
#include "../../../../chrome/browser/browser_process_impl.h"
#undef Init
```
Would it be possible to do something similar in `chromium_src/components/sync/driver/profile_sync_service.h` and get rid of this patch?

Reply:
Double checked. Here is diff of my attempt
```
 git diff chromium_src/components/sync/driver/profile_sync_service.h 
diff --git a/chromium_src/components/sync/driver/profile_sync_service.h b/chromium_src/components/sync/driver/profile_sync_service.h
index a1506da313..6c890903e4 100644
--- a/chromium_src/components/sync/driver/profile_sync_service.h
+++ b/chromium_src/components/sync/driver/profile_sync_service.h
@@ -10,8 +10,12 @@
  private:                             \
   friend class BraveProfileSyncService;
 
+#define Initialize virtual Initialize
+
 #include "../../../../../components/sync/driver/profile_sync_service.h"
 
+#undef Initialize
+
 #undef BRAVE_PROFILE_SYNC_SERVICE_H_
 
 #endif  // BRAVE_CHROMIUM_SRC_COMPONENTS_SYNC_DRIVER_PROFILE_SYNC_SERVICE_H_
```
Here is the error, it tries to apply macro to 
`static inline void Initialize(TypeOnMemory* value, Arena* arena); ` at `third_party/protobuf/src/google/protobuf/map_type_handler.h:160` :
```
[10/4185] CXX obj/brave/components/brave_sync/profile_sync_service_helper/profile_sync_service_helper.o
FAILED: obj/brave/components/brave_sync/profile_sync_service_helper/profile_sync_service_helper.o 
/mnt/EVO970/Projects/br_br_c89/brave-browser/src/brave/script/redirect-cc.py ../../third_party/llvm-build/Release+Asserts/bin/clang++ -MMD -MF obj/brave/components/brave_sync/profile_sync_service_helper/profile_sync_service_helper.o.d -DBRAVE_CHROMIUM_BUILD -DDCHECK_ALWAYS_ON=1 -DUSE_UDEV -DUSE_AURA=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_OZONE=1 -DUSE_X11=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_GNU_SOURCE -DCR_CLANG_REVISION=\"llvmorg-12-init-15676-g5c951623-1\" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DCOMPONENT_BUILD -D_LIBCPP_ABI_UNSTABLE -D_LIBCPP_ABI_VERSION=Cr -D_LIBCPP_ENABLE_NODISCARD -D_LIBCPP_DEBUG=0 -DCR_LIBCXX_REVISION=375504 -DCR_SYSROOT_HASH=22f2db7711f7426a364617bb6d78686cce09a8f9 -D_DEBUG -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_40 -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_40 -DWEBP_EXTERN=extern -DVK_USE_PLATFORM_XCB_KHR -DGL_GLEXT_PROTOTYPES -DUSE_GLX -DUSE_EGL -DABSL_CONSUME_DLL -DBORINGSSL_SHARED_LIBRARY -DGOOGLE_PROTOBUF_NO_RTTI -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER -DHAVE_PTHREAD -DPROTOBUF_USE_DLLS -DLEVELDB_PLATFORM_CHROMIUM=1 -DLEVELDB_PLATFORM_CHROMIUM=1 -DLEVELDB_SHARED_LIBRARY -I../../brave/chromium_src -I../.. -Igen -I../../third_party/perfetto/include -Igen/third_party/perfetto/build_config -Igen/third_party/perfetto -I../../third_party/libyuv/include -I../../third_party/jsoncpp/source/include -I../../third_party/libwebp/src -I../../third_party/vulkan-deps/vulkan-headers/src/include -I../../third_party/khronos -I../../gpu -Igen/third_party/dawn/src/include -I../../third_party/dawn/src/include -I../../third_party/abseil-cpp -I../../third_party/boringssl/src/include -I../../third_party/protobuf/src -Igen/protoc_out -Igen -I../../third_party/protobuf/src -Igen/third_party/metrics_proto -Igen/net/third_party/quiche/src -I../../net/third_party/quiche/src -I../../third_party/leveldatabase -I../../third_party/leveldatabase/src -I../../third_party/leveldatabase/src/include -fno-delete-null-pointer-checks -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -funwind-tables -fPIC -pthread -fcolor-diagnostics -fmerge-all-constants -fcrash-diagnostics-dir=../../tools/clang/crashreports -mllvm -instcombine-lower-dbg-declare=0 -fcomplete-member-pointers -m64 -march=x86-64 -msse3 -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -Xclang -fdebug-compilation-dir -Xclang . -no-canonical-prefixes -Wall -Werror -Wextra -Wimplicit-fallthrough -Wthread-safety -Wextra-semi -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-psabi -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-final-dtor-non-final-class -Wno-builtin-assume-aligned-alignment -Wno-deprecated-copy -Wno-non-c-typedef-for-linkage -Wmax-tokens -O0 -fno-omit-frame-pointer -g2 -Xclang -debug-info-kind=constructor -ggnu-pubnames -ftrivial-auto-var-init=pattern -fvisibility=hidden -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang check-ipc -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -isystem../../build/linux/debian_sid_amd64-sysroot/usr/include/glib-2.0 -isystem../../build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu/glib-2.0/include -DPROTOBUF_ALLOW_DEPRECATED=1 -Wno-shorten-64-to-32 -Wno-header-guard -isystem../../build/linux/debian_sid_amd64-sysroot/usr/include/nss -isystem../../build/linux/debian_sid_amd64-sysroot/usr/include/nspr -Wno-undefined-bool-conversion -Wno-tautological-undefined-compare -std=c++14 -fno-trigraphs -Wno-trigraphs -fno-exceptions -fno-rtti -nostdinc++ -isystem../../buildtools/third_party/libc++/trunk/include -isystem../../buildtools/third_party/libc++abi/trunk/include --sysroot=../../build/linux/debian_sid_amd64-sysroot -fvisibility-inlines-hidden -Wno-deprecated-declarations -c ../../brave/components/brave_sync/profile_sync_service_helper.cc -o obj/brave/components/brave_sync/profile_sync_service_helper/profile_sync_service_helper.o
In file included from ../../brave/components/brave_sync/profile_sync_service_helper.cc:11:
In file included from ../../brave/components/sync/driver/brave_sync_profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:160:22: error: 'virtual' can only appear on non-static member functions
  static inline void Initialize(Type** x, Arena* arena);
                     ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../brave/components/brave_sync/profile_sync_service_helper.cc:11:
In file included from ../../brave/components/sync/driver/brave_sync_profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:214:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(STRING)
^
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:210:24: note: expanded from macro 'MAP_HANDLER'
    static inline void Initialize(TypeOnMemory* value, Arena* arena);         \
                       ^
../../brave/chromium_src/components/sync/driver/profile_sync_service.h:13:20: note: expanded from macro 'Initialize'
#define Initialize virtual Initialize
                   ^
In file included from ../../brave/components/brave_sync/profile_sync_service_helper.cc:11:
In file included from ../../brave/components/sync/driver/brave_sync_profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/profile_sync_service.h:15:
In file included from ../../brave/chromium_src/components/sync/driver/../../../../../components/sync/driver/profile_sync_service.h:27:
In file included from ../../components/sync/base/sync_prefs.h:25:
In file included from gen/components/sync/protocol/sync.pb.h:26:
In file included from ../../third_party/protobuf/src/google/protobuf/generated_message_table_driven.h:34:
In file included from ../../third_party/protobuf/src/google/protobuf/map.h:56:
../../third_party/protobuf/src/google/protobuf/map_type_handler.h:215:1: error: 'virtual' can only appear on non-static member functions
MAP_HANDLER(BYTES)..................
```

Reply:
Ah! I see it now, the problem is happening because there's another `Initialize()` methods in another header that gets transitively included via some of `profile_sync_service.h`'s own headers.

In this case, I believe a patch as you have is probably the best option, although I think the recommended way is to simply add the `virtual` in a separate line. That is, instead of this:
```diff
-  void Initialize();
+  void virtual Initialize();
```

...maybe something like this:
```diff
+  virtual
    void Initialize();
```

[See `patches/chrome-browser-ui-views-toolbar-toolbar_view.h.patch`](https://github.com/brave/brave-core/blob/master/patches/chrome-browser-ui-views-toolbar-toolbar_view.h.patch) for instance

Reply:
@mariospr , oh indeed, thanks, I'll update

Reply:
Updated patch at https://github.com/brave/brave-core/pull/7640/files#diff-3a5ca35846ae4143638fb9aec61b7dec092ee538823e6eec4c60192c2856257bR9

Reply:
@AlexeyBarabash @mariospr How about adding some headers that have `Initialize()` like below?
With this, we can avoid patching :)
``` #ifndef BRAVE_CHROMIUM_SRC_COMPONENTS_SYNC_DRIVER_PROFILE_SYNC_SERVICE_H_
 #define BRAVE_CHROMIUM_SRC_COMPONENTS_SYNC_DRIVER_PROFILE_SYNC_SERVICE_H_

+// Included because these headers also have |Initialize() method.
+#include "components/sync/engine/sync_engine.h"
+#include "third_party/protobuf/src/google/protobuf/generated_message_table_driven.h"
+
 #define BRAVE_PROFILE_SYNC_SERVICE_H_ \
  private:                             \
   friend class BraveProfileSyncService;

+#define Initialize virtual Initialize
+
 #include "../../../../../components/sync/driver/profile_sync_service.h"

 #undef BRAVE_PROFILE_SYNC_SERVICE_H_
+#undef Initialize

 #endif  // BRAVE_CHROMIUM_SRC_COMPONENTS_SYNC_DRIVER_PROFILE_SYNC_SERVICE_H_
```

Reply:
@simonhong , thanks this works, updated the PR at
https://github.com/brave/brave-core/pull/7640/commits/d02871f76a65a556fb6fcc177e07e1e082d2df75#diff-b39f0315cc67357acd8fed994e88ae553358b871f9ae004950d83fc8d1211b73R19